### PR TITLE
Replace lazy `ensureInitialized` patterns with explicit initialization

### PR DESCRIPTION
**Pull request by ZwFink**
This PR replaces lazy initialization in `MPIStorageCache` with eager.
Importantly, this fixes an `MPIStorageCache` edge case where Rank 0 never starts the comm thread because it doesn't access the storage cache.
As a consequence, `proteus::init` is now a _required call_ for cache initialization.
Therefore, I have updated tests to use `proteus::init/finalize`.

I have decoupled the `ObjectCacheChain` and `JitEngine` to ensure that `JitEngine` construction remains lazy. If `JitEngine` is eagerly constructed in `proteus::init`, then a bug is introduced in one particular case. If proteus is built with HIP/CUDA enabled, but no device is visible, eager GPU `JitEngine` construction fails in HIP/CUDA runtime calls. Hence, the `ObjectCacheRegistry` singleton to decouple `JitEngine` and `ObjectCacheChain`. Such interconnected global state seems a code smell to me, but I'm not sure how to fix it.

### CODE REVIEW

File: src/runtime/Frontend/Dispatcher.cpp [hunk 1]
```diff
@@ -67,6 +68,13 @@ Dispatcher &Dispatcher::getDispatcher(TargetModelType TargetModel) {
   }
 }
 
+ObjectCacheChain &Dispatcher::getObjectCache() {
+  auto CacheOpt = ObjectCacheRegistry::instance().get(DispatcherName);
+  if (!CacheOpt)
+    reportFatalError("ObjectCache missing for " + DispatcherName + ".");
+  return CacheOpt->get();
+}
+
```
[the remark this answers is not in the document]
Comment (ZwFink) [reply src/runtime/Frontend/Dispatcher.cpp]: I have realized the `Registry` is not needed. I have replaced it with a `unique_ptr` in `JitEngine`s that is populated in `init` calls. Therefore, we don't have this awkward indirection.
